### PR TITLE
Hierarchical clustering

### PR DESCRIPTION
**Pull request by mawada-sweis**
I recommend **four clusters** as the number of clusters with the greatest clustering *after applying PCA*.

### CODE REVIEW

File: src/utils/hierarchical_model.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    plt.show()
+
+
+def plot_clustering_after_reduction(labels: ndarray, transformed: ndarray, dimensions: int) -> None:
+    """Generates a scatter plot of the clustering of data after reducing using the given transformed data by PCA.
+
+    Args:
+        data (pd.DataFrame): The input dataset to be plotted.
+        transformed (ndarray): The transformed data using PCA. 
+        dimensions (int): The number of dimensions to reduce the data to after plotting.
```
Comment (Rama-Has) [inline src/utils/hierarchical_model.py:185]: I think this description is misleading did you mean number of components generated by the PCA?

File: src/utils/hierarchical_model.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        dimensions (int): The number of dimensions to keep after the transformation.
+    Returns:
+        ndarray: A numpy ndarray of shape (n_samples, dimensions) representing the transformed data.
+    """
+    # Perform PCA on the data
+    pca = PCA(n_components=dimensions)
+    return pca.fit_transform(data)
+
+
+def plot_clustering_before_reduction(data: pd.DataFrame, labels:ndarray, dimensions: int) -> None:
```
Comment (Rama-Has) [inline src/utils/hierarchical_model.py:139]: Maybe you can apply some minor changes to this function to get rid of plot_clustering_after_reduction().
Minor changes such as adding an optional argument to check if the data need a dimensionality reduction before plotting or not.

File: src/utils/hierarchical_model.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    Returns:
+        ndarray: An array of length n. T[i] is the flat cluster number to which original observation i belongs.
+    """
+    return fclusterdata(data, t=k, criterion='maxclust', method='ward')
+
+
+def _map_color_cluster(labels: ndarray) -> pd.Series(str):
+    """Assigns a color to each cluster in the given DataFrame based on their label.
+    Args:
+        df (pd.DataFrame): The DataFrame containing the cluster labels.
```
Comment (Rama-Has) [inline src/utils/hierarchical_model.py:109]: There is no dataframe, u did not attach "labels" to the documentation.

File: src/utils/hierarchical_model.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    plt.title("Dendrogram with Threshold")
+    
+    # Draw the horizontal line
+    plt.axhline(y=distance_threshold, color='black', linestyle='--')
+    
+    # Show the plot
+    plt.show()
+
+
+def hierarchical_clustering(data: pd.DataFrame, k=3) -> ndarray:
```
Comment (Rama-Has) [inline src/utils/hierarchical_model.py:92]: No need for this function.

**Review by Rama-Has — COMMENTED** [review]
**Please add an evaluation method to select best k rather than selecting it visually.**

**Comment by mawada-sweis** [comment]
> **Please use an evaluation method to select best k rather than selecting it visually.**

I display the silhouette score for each one...

File: src/utils/hierarchical_model.py (hunk 3 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py:109]: I use this function in several places in the notebooks.

Comment (mawada-sweis) [reply src/utils/hierarchical_model.py:109]: I fixed it

File: src/utils/hierarchical_model.py (hunk 4 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py:92]: I use this function in several places in the notebooks.

File: src/utils/hierarchical_model.py (hunk 2 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py]: Thank you, I said I'd do it but then forgot... I grouped them under the function name "plot clustering".

File: src/utils/hierarchical_model.py (hunk 3 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py]: Fixed it.

File: src/utils/hierarchical_model.py (hunk 1 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py]: I update the documentations of 'hierarchical_model' file.

File: src/utils/hierarchical_model.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    Args:
+        data (pandas.DataFrame): The dataset to cluster.
+        method_type (str, optional): The clustering algorithm to use. Default is 'ward'.
+        distance_metric (str, optional): The distance metric to use. Default is 'euclidean'.
+        optimal_ordering (bool, optional): Whether to compute the optimal leaf ordering for the dendrogram. Default is False.
+
+    Returns:
+        numpy.ndarray: The hierarchical clustering linkage matrix.
+    """
+    return linkage(data, method=method_type, metric=distance_metric, optimal_ordering=optimal_ordering)
```
Comment (zubaidasader) [inline src/utils/hierarchical_model.py:34]: no need for a function to do this

File: src/utils/hierarchical_model.py (hunk 4 above)
Comment (zubaidasader) [reply src/utils/hierarchical_model.py:92]: Yes but the function only returns a value. you can do this without initializing a function so why did you use it?

**Review by zubaidasader — APPROVED** [review]

**Review by eleenkmail — APPROVED** [review]

**Review by Rama-Has — APPROVED** [review]

File: src/utils/hierarchical_model.py (hunk 5 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py:34]: I use this function in several places in the notebooks.

**Comment by mawada-sweis** [comment]
Because the standard scaler is not greater than minmax scaler as we discuss at the meeting, so I don't use it.

**Comment by mawada-sweis** [comment]
> _zubaidasader commented on 2023-03-04T21:14:53Z_ ----------------------------------------------------------------
> 
> These results are on the MinMax scaled data. why you didn't use the Standard Scaled data?

Because the standard scaler is not greater than minmax scaler as we discuss at the meeting, so I don't use it.

File: src/utils/hierarchical_model.py (hunk 4 above)
Comment (mawada-sweis) [reply src/utils/hierarchical_model.py:92]: It's easier...
